design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,478 @@
+agnostic. The purpose of this document is to explore whether the OAM
+specification could be made friendlier to - idiomatic to and compatible with -
+Kubernetes powered runtimes while remaining flexible enough to be usefully
+implemented by other runtimes.
+
+Kubernetes powered runtimes are the focus of this document because the OAM
+specification already looks and feels a lot like the Kubernetes API. When it is
+implemented using the _actual_ Kubernetes API any divergences from said API’s
+conventions are an order of magnitude more likely to violate the [principle of
+least astonishment] for users than for users of other runtimes.
+
+Any proposal outlined by this document should ensure:
+
+* The separation of concerns that is the core focus of the OAM specification.
+* That the OAM specification “look and feel” idiomatic to the Kubernetes API
+  model.
+* That the OAM specification remains useful and compelling to other runtimes.
+
+Note that this document frequently uses Crossplane as an example of a tentative
+OAM runtime that would benefit from a Kubernetes friendly OAM specification, but
+defining specifically how Crossplane will support OAM is out of scope.
+
+## Proposal
+
+This document proposes:
+
+1. An ApplicationConfiguration _references_ distinct Components and Trait
+   resources, grouping them into one logical application and providing inputs to
+   their parameters. This enables the complete schema for an
+   ApplicationConfiguration to be expressed in one place (one
+   CustomResourceDefinition in Kubernetes runtimes), rather than many.
+1. Each Component embed an OAM “workload” under the `.spec.workload` key.
+   Available workloads are defined by WorkloadDefinitions. In Kubernetes
+   runtimes any existing custom resource with a corresponding WorkloadDefinition
+   can be used in a Component.
+1. The set of available Traits and their schemas be defined via
+   TraitDefinitions.
+1. The set of available Scopes and their schemas be defined via
+   ScopeDefinitions.
+
+### Components and Workloads
+
+Each Component is of a distinct workload _kind_. This is functionally equivalent
+to the existing concept of workload _types_, but reuses the _kind_ term already
+used within Kubernetes and OAM. The schema of a workload kind - including core
+workload kinds like Server - are defined via a WorkloadDefinition. For example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: WorkloadDefinition
+metadata:
+  name: server.core.oam.dev
+spec:
+  definitionRef:
+    name: server.core.oam.dev
+```
+
+The above WorkloadDefinition references a Kubernetes CustomResourceDefinition
+named server.core.oam.dev, designating custom resources defined by said CRD to
+be valid OAM workloads. OAM runtimes that are not powered by Kubernetes could
+opt to either implement the CRD concept, or introduce a new field to the
+WorkloadDefinition spec that defines a kind of workload and its schema.
```

Comment:
Hi Nic, Nice proposal!  I can see there are a lot of thoughts in it.   The unique features of K8N are leveraged and I can see it married well with OAM schema.  
	
I am kind of confused in this section.  Hope that you can help.
	
I believe at one point, we are saying the "Core" schema will be enforced; while the "Extended" will be open up for different providers.  This seen to be the Core Server workload definition, but what does it means by "introduce a new field to the WorkloadDefinition spec that defines a kind of workload and its schema"?  Does it means Developers or Application Operators need to specify parameter based on the defined schema?

Reply:
> what does it means by "introduce a new field to the WorkloadDefinition spec that defines a kind of workload and its schema"? Does it means Developers or Application Operators need to specify parameter based on the defined schema?

The goal here was to point out that OAM runtimes that are not powered by Kubernetes could choose to add more settings related to workloads by adding new optional fields to the spec of a `WorkloadDefinition` (for example a field that defined an inline schema instead of pointing to a `definitionRef`). I don't expect application developers or operators will need to interact with workload definitions in any way apart from browsing them to discover which workloads are available and what their parameters are. Does that make sense?

Reply:
Hi Nic, thanks for the clarification.
From how I understand, Component become instance of Workload.  
If we add settings to a workload, we are expecting developer or operator to set parameters to those settings.
In this case, the settings schema of Core Workloads become a concern.  
In the WorkloadDefinition example above, "server.core.oam.dev" is a CRD and it is definitionRef.  
Am I correct that the settings of the parameters, for the Component example below, comes from "server.core.oam.dev"?
Also, "a field that defined an inline schema instead of pointing to a definitionRef".  
These sound like we are allowing workload implementations to define their own setting schema.
Is this correct?

Reply:
We discussed and clarified this in this morning's community meeting. For posterity, I believe @abchow's primary concern was that core workload kinds would have identical schemas across OAM runtimes; i.e. you would configure a `server.oam.dev/v1alpha` workload kind exactly the same on a Kubernetes OAM runtime as you would on an AcmeCloud OAM runtime. I believe this would be the case.

Reply:
Thanks Nic.  For clarification, my concern was the Settings / Parameters of Core types are not enforced.  If it is not enforced, different OAM runtime providers can provide their own setting attributes.  For example, Provider A can call container image as "imageLoc"; while Provider B can call it "swrImage".  In this case, the model that work in Provider A will not work on Provider B without modification, and it defeated the goal of Platform-Independent.
It was make clear to me that this proposal will not affect the Settings / Parameters list of the Core types.  Thanks